Human readable dates and multiple file path support in the collate command

rhyshort opened the pull request:
- [x] Tick to sign-off your agreement to the [Developer Certificate of Origin (DCO) 1.1](../blob/master/DCO1.1.txt)


## What

- Make `--start` and `--end` options in collate accept humna readable dates, ie 2022-09-29 instead of 20220929
- Enable collation of mulitple files with a single collate command by using `,` as a path seperator.

## Why

The current `--start` and `--end` formats are hard for humans to read, which makes collating data to provide to auditors difficult, as conversion between date formats needs to be done by hand, and the number grouping makes it hard to ensure there have not been any typos etc

Being able to pull multiple files at once makes it easier to pull multiple files over the same date range when an auditor is requesting multiple files.

## How

- Replace all references to `YYYYMMDD` dates to `YYYY-MM-DD` dates
- Split the file path string and iterate over the result calling collate functions multiple times.

## Test

- Updated tests for the new date format
- Add a new test to pull multiple files.

## Context

rhyshort committed:
Use human friendly date format in collate

Use human friendly date format in collate, so when collating a large
number of files becomes easy to translate requested dates from auditors
into cli arguments. There is compatbility between the new human friendly
format and the previous format negate the need for a breaking change
release and avoid users changing their scripts.

drsm79 commented on harvest/cli.py:
```diff
@@ -90,7 +90,8 @@ def _init_arguments(self):
             'filepath',
             help=(
                 'the relative path to a file in a git repository '
```
I think better would be to use an `append` action, so the list is explicit

drsm79 replied:
https://docs.python.org/3/library/argparse.html

rhyshort replied:
`append` action wasn't doing quite the right thing with postional args, so option to use `nargs='+'`, which has the desired behaviour. Example help text: 
```
harvest collate --help
usage: harvest collate [-h] [--repo-path ~/path/git-repo] [--creds ~/path/creds] [--end YYYY-MM-DD] [--start YYYY-MM-DD] [--version] repo filepath [filepath ...]

positional arguments:
  repo                  the URL to the repository containing files to be processed by harvest, as an example: https://github.com/my-org/my-repo or: local - if working exclusively with a local git repo
  filepath              the relative path to a file in a git repository that you wish to retrieve.

optional arguments:
  -h, --help            show this help message and exit
  --repo-path ~/path/git-repo
                        the operating system location of a local git repository - if not provided, repo path is assumed to be $TMPDIR/harvest
  --creds ~/path/creds  the path to credentials file - defaults to ~/.credentials
  --end YYYY-MM-DD      the end of date range for the file you wish to retrieve - defaults to the current date
  --start YYYY-MM-DD    the start of date range for the file you wish to retrieve - defaults to same value as the end of date range
  --version             the harvest version```

drsm79 commented on harvest/cli.py:
```diff
@@ -90,16 +90,17 @@ def _init_arguments(self):
                 'defaults to the current date'
             ),
-            metavar='YYYYMMDD',
+            metavar='YYYY-MM-DD',
```
Sure I mentioned that this is an unnecessary breaking change. Suggest we have a function that takes a date in either YYYYMMDD or YYYY-MM-DD format and returns the datetime object for it. That will preserve current behaviour for anyone who's got a script using the tool and support both ISO date formats.

rhyshort replied:
took me some time to get around to it but done now

drsm79 replied:
Nice one

rhyshort commented:
hmm the github action looks like is failing because of the python version is too old for the version of flake8.

cletomartin commented:
@rhyshort I'm changing all the linting side of the project here #28

rhyshort commented:
@cletomartin nice, I'll rebase this on top of #28 once merged.

rhyshort committed:
Collate multiple files for the same date range

Using comma seperated paths, enable collate to pull multiole files at
once. This makes it easier for humans to pull multiple bits of evidence
at once

rhyshort commented:
Rebased on top of the changes from #28

cletomartin reviewed (commented):
Thanks!

Please update the version https://github.com/ComplianceAsCode/auditree-harvest/blob/main/harvest/__init__.py#L16 and the CHANGES.md file too.

rhyshort committed:
Update version number and Changes

Update version to 1.3.0 and add changes entry.

rhyshort commented:
Done

rhyshort commented:
Thanks @cletomartin, I don't have write access so can you go ahead and merge please?

drsm79 commented:
Merged - thanks Rhys!
